kafka/signed_token_redeem_handler.go

```diff
@@ -10,50 +10,82 @@ import (
+/*
+ SignedTokenRedeemHandler emits payment tokens that correspond to the signed confirmation
+ tokens provided. If it encounters a permanent error, it emits a permanent result for that
+ item. If the error is temporary, an error is returned to indicate that progress cannot be
+ made.
+*/
 func SignedTokenRedeemHandler(
-	data []byte,
+	msg kafka.Message,
 	producer *kafka.Writer,
 	server *cbpServer.Server,
-	logger *zerolog.Logger,
+	log *zerolog.Logger,
 ) error {
-	const (
-		redeemOk                  = 0
-		redeemDuplicateRedemption = 1
-		redeemUnverified          = 2
-		redeemError               = 3
-	)
+	data := msg.Value
+	// Deserialize request into usable struct
 	tokenRedeemRequestSet, err := avroSchema.DeserializeRedeemRequestSet(bytes.NewReader(data))
 	if err != nil {
-		return fmt.Errorf("request %s: failed avro deserialization: %w", tokenRedeemRequestSet.Request_id, err)
+		message := fmt.Sprintf("request %s: failed avro deserialization", tokenRedeemRequestSet.Request_id)
+		handlePermanentRedemptionError(
+			message,
+			msg,
+			producer,
+			tokenRedeemRequestSet.Request_id,
+			int32(avroSchema.RedeemResultStatusError),
+			log,
+		)
+		return nil
 	}
-	defer func() {
-		if recover() != nil {
-			logger.Error().
-				Err(fmt.Errorf("request %s: redeem attempt panicked", tokenRedeemRequestSet.Request_id)).
-				Msg("signed token redeem handler")
-		}
-	}()
+
+	logger := log.With().Str("request_id", tokenRedeemRequestSet.Request_id).Logger()
+
 	var redeemedTokenResults []avroSchema.RedeemResult
+	// For the time being, we are only accepting one message at a time in this data set.
+	// Therefore, we will error if more than a single message is present in the message.
 	if len(tokenRedeemRequestSet.Data) > 1 {
 		// NOTE: When we start supporting multiple requests we will need to review
 		// errors and return values as well.
-		return fmt.Errorf("request %s: data array unexpectedly contained more than a single message. this array is intended to make future extension easier, but no more than a single value is currently expected", tokenRedeemRequestSet.Request_id)
+		message := fmt.Sprintf("request %s: data array unexpectedly contained more than a single message. This array is intended to make future extension easier, but no more than a single value is currently expected", tokenRedeemRequestSet.Request_id)
+		handlePermanentRedemptionError(
+			message,
+			msg,
+			producer,
+			tokenRedeemRequestSet.Request_id,
+			int32(avroSchema.RedeemResultStatusError),
+			log,
+		)
+		return nil
 	}
 	issuers, err := server.FetchAllIssuers()
 	if err != nil {
-		return fmt.Errorf("request %s: failed to fetch all issuers: %w", tokenRedeemRequestSet.Request_id, err)
+		if processingError, ok := err.(*utils.ProcessingError); ok && processingError.Temporary {
+			return processingError
+		}
+		message := fmt.Sprintf("request %s: failed to fetch all issuers", tokenRedeemRequestSet.Request_id)
+		handlePermanentRedemptionError(
```

Comment:
are there non-temporary failures of `FetchAllIssuers`? I'm struggling to think of any